libtcmu_log.c

```diff
@@ -113,14 +216,161 @@ void tcmu_info_message(const char *funcname, int linenr, const char *fmt, ...)
 	va_list args;
 
 	va_start(args, fmt);
-	tcmu_log_internal(TCMU_LOG_INFO, funcname, linenr, fmt, args);
+	log_internal(TCMU_LOG_INFO, funcname, linenr, fmt, args);
 	va_end(args);
 }
+
 void tcmu_dbg_message(const char *funcname, int linenr, const char *fmt, ...)
 {
 	va_list args;
 
 	va_start(args, fmt);
-	tcmu_log_internal(TCMU_LOG_DEBUG, funcname, linenr, fmt, args);
+	log_internal(TCMU_LOG_DEBUG, funcname, linenr, fmt, args);
 	va_end(args);
 }
+
+static void log_output(int pri, const char *msg)
+{
+	log_to_syslog(pri, msg);
+	//log to stdout if tcmu_log_level is DEBUG
+	if (pri <= TCMU_LOG_DEBUG && tcmu_log_level == TCMU_LOG_DEBUG) {
+		fprintf(stdout, "%s", msg);
+	}
+}
+
+static bool log_buf_not_empty_output(struct log_buf *logbuf)
+{	
+	unsigned int tail;
+	uint8_t pri;
+	char *msg, buf[LOG_MSG_LEN];
+	
+	if (!logbuf) {
+		return false;
+	}
+
+	pthread_mutex_lock(&logbuf->lock);
+	if (rb_is_empty(logbuf)) {
+		pthread_mutex_unlock(&logbuf->lock);
+		return false;
+	}
+
+	tail = logbuf->tail;
+	pri = rb_get_pri(logbuf, tail);
+	msg = rb_get_msg(logbuf, tail);
+	memcpy(buf, msg, LOG_MSG_LEN);
+	rb_update_tail(logbuf);
+	pthread_mutex_unlock(&logbuf->lock);
+
+	/*
+ 	 * This may block due to rsyslog and syslog-ng, etc.
+ 	 * And the log productors could still insert their log
+ 	 * messages into the ring buffer without blocking. But
+ 	 * the ring buffer may lose some old log rbs if the
+ 	 * ring buffer is full.
+ 	 */
+	log_output(pri, buf);
+
+	return true;
+}
+
+static void cancel_log_thread(pthread_t thread)
+{
+	void *join_retval;
+	int ret;
+
+	ret = pthread_cancel(thread);
+	if (ret) {
+		return;
+	}
+
+	ret = pthread_join(thread, &join_retval);
+	if (ret) {
+		return;
+	}
+
+	if (join_retval != PTHREAD_CANCELED)
+		return;
+}
+
+void tcmu_cancel_log_thread(void)
+{
+	cancel_log_thread(logbuf->thread_id);
+}
+
+static void log_thread_cleanup(void *arg)
+{
+	struct log_buf *logbuf = arg;
+
+	pthread_cond_destroy(&logbuf->cond);
+	pthread_mutex_destroy(&logbuf->lock);
+	free(logbuf);
+
+	close_syslog();
+}
+
+static void *log_thread_start(void *arg)
+{
+	struct log_buf *logbuf = arg;
+
+	pthread_cleanup_push(log_thread_cleanup, arg);
+
+	open_syslog(NULL, 0, 0);
+
+	while (1) {
+		pthread_mutex_lock(&logbuf->lock);
+		if (!logbuf->finish_initialize) {
+			logbuf->finish_initialize = true;
+			pthread_cond_signal(&logbuf->cond);
```

Comment:
Maybe move this chunk before the while since it is only needed for the initial sync up.